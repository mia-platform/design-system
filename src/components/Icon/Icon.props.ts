/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import * as AntIcons from 'react-icons/ai'
import * as FeatherIcons from 'react-icons/fi'
import * as PhosphorIcons from 'react-icons/pi'

import MiaPlatform from './assets/MiaPlatform.svg'
import MiaPlatformColored from './assets/MiaPlatformColored.svg'

/**
 * Custom icons for brand images.
 *
 * @remarks This object should only be used for brand image-related custom SVGs.
 */
export const customIcons = {
  MiaPlatform,
  MiaPlatformColored,
} as const

/**
 * Supported icon packs.
 *
 * @see {@link https://github.com/feathericons/feather Feather Icons}
 * @see {@link https://github.com/phosphor-icons/core Phosphor Icons}
 * @see {@link https://github.com/ant-design/ant-design-icons Ant Icons}
 */
export const reactIcons = {
  ...FeatherIcons,
  ...PhosphorIcons,
  ...AntIcons,
} as const

export type IconProps = {

  /**
   * The color of the icon.
   */
  color?: string,

  /**
   * The name of the icon.
   */
  name: keyof typeof customIcons | keyof typeof reactIcons,

  /**
   * The size of the icon.
   */
  size?: 16 | 24 | 32 | 48 | 64 | 96,
}
