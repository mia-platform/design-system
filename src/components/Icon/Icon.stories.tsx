import type { Meta, StoryObj } from '@storybook/react'

import { Icon } from '.'

const iconsUrl = 'https://react-icons.github.io/react-icons/search'

const meta = {
  component: Icon,
  parameters: {
    design: { type: 'iframe', name: 'Search', url: iconsUrl },
  },
  args: {
    ...Icon.defaultProps,
  },
} satisfies Meta<typeof Icon>

export default meta
type Story = StoryObj<typeof meta>

export const Custom: Story = {
  args: { ...meta.args, name: 'MiaPlatform' },
}

export const Ant: Story = {
  args: { ...meta.args, name: 'AiOutlineHome' },
}

export const Feather: Story = {
  args: { ...meta.args, name: 'FiHome' },
}

export const Phosphor: Story = {
  args: { ...meta.args, name: 'PiHouse' },
}

