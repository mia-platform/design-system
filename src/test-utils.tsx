/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { RenderOptions, RenderResult, render } from '@testing-library/react'
import { ConfigProvider } from 'antd'
import { ReactNode } from 'react'

import { ThemeContext, defaultTheme } from './components/ThemeProvider'
import { ANT_PREFIX } from './components/ThemeProvider/Ant'

/**
 * Wraps a React node with a series of providers for testing purposes.
 *
 * @remarks It adds the default theme to components
 * @remarks It adds the default Ant class prefix
 * @remarks It removes the css-do-not-override classes from Ant Design components
 *
 * @param {ReactNode} children - The React node to be wrapped.
 * @returns {ReactNode} The wrapped React node.
 */
const TestWrapper = ({ children }: {children: ReactNode}): ReactNode => (
  <ThemeContext.Provider value={defaultTheme}>
    <ConfigProvider prefixCls={ANT_PREFIX} theme={{ hashed: false }}>
      {children}
    </ConfigProvider>
  </ThemeContext.Provider>
)

/**
 * Custom render function for testing React components.
 *
 * @param {ReactNode} element - The React node to be rendered.
 * @param {RenderOptions} [options] - Options for rendering.
 * @returns {RenderResult} The result of rendering the React node.
 */
const customRender = (
  element: ReactNode,
  options?: Omit<RenderOptions, 'wrapper'>,
): RenderResult => render(element, { wrapper: TestWrapper, ...options })

export * from '@testing-library/react'
export {
  customRender as render,
  TestWrapper,
}
