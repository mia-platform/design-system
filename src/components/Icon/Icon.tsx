/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ReactElement, useContext } from 'react'
import { IconContext } from 'react-icons'

import { IconProps, customIcons, reactIcons } from './Icon.props'
import log from '../../utils/log'

/**
 * UI component for displaying different icon packs (Ant, Feather, Phosphor) and custom SVGs
 *
 * @link https://react-icons.github.io/react-icons/
 * @returns {Icon} Icon component
 */
export const Icon = ({
  name,
  size,
  color,
}: IconProps): ReactElement | null => {
  const { color: defaultColor, size: defaultSize, className } = useContext(IconContext)

  const IconComponent = name in customIcons
    ? customIcons?.[name as keyof typeof customIcons]
    : reactIcons?.[name as keyof typeof reactIcons]

  if (!IconComponent) {
    log.error(`icon name ${name} not supported`)
    return null
  }

  return (
    <IconComponent
      alt={name}
      aria-label={name}
      className={className}
      color={color ?? defaultColor}
      height={size ?? defaultSize}
      role={'img'}
      size={size ?? defaultSize}
      width={size ?? defaultSize}
    />
  )
}

Icon.defaultProps = {
  size: 24 as const,
}
