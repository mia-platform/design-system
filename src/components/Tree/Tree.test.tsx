/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { TreeDataNode } from 'antd'

import { fireEvent, render, screen, waitFor } from '../../test-utils'
import { Icon } from '../Icon'
import { Tree } from '.'

const defaultTreeData: TreeDataNode[] = [
  {
    title: 'Fruits',
    key: 'fruits',
    children: [
      {
        title: 'Favorites',
        key: 'favorites',
        children: [
          {
            title: 'Apple',
            key: 'apple',
          },
          {
            title: 'Banana',
            key: 'banana',
          },
          {
            title: 'Peach',
            key: 'peach',
          },
        ],
      },
      {
        title: 'Others',
        key: 'others',
        children: [
          {
            title: 'Strawberry',
            key: 'strawberry',
          },
          {
            title: 'Pineapple',
            key: 'pineapple',
          },
          {
            title: 'Fig',
            key: 'fig',
          },
        ],
      },
    ],
  },
  {
    title: 'Vegetables',
    key: 'vegetables',
    children: [
      {
        title: 'Spinach',
        key: 'spinach',
      },
      {
        title: 'Leek',
        key: 'leek',
      },
    ],
  },
]

describe('Tree Component', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  test('renders Tree', async() => {
    const { asFragment } = render(<Tree treeData={defaultTreeData} />)

    expect(asFragment()).toMatchSnapshot()

    expect(screen.getByText('Vegetables')).toBeInTheDocument()
    const [, vegetablesCaret] = screen.getAllByRole('img', { name: /caret-down/i })
    fireEvent.click(vegetablesCaret)

    await waitFor(() => { expect(screen.getByText('Spinach')).toBeInTheDocument() })
    expect(screen.getByText('Leek')).toBeInTheDocument()
  })

  test('nodes can be selected', async() => {
    const onSelectMock = jest.fn()
    render(<Tree treeData={defaultTreeData} onSelect={onSelectMock} />)

    const [, vegetablesCaret] = screen.getAllByRole('img', { name: /caret-down/i })
    fireEvent.click(vegetablesCaret)

    const leekNode = await screen.findByText('Leek')
    fireEvent.click(leekNode)

    await waitFor(() => { expect(onSelectMock).toHaveBeenCalledTimes(1) })
    expect(onSelectMock).toHaveBeenCalledWith(['leek'], expect.any(Object))
  })

  test('nodes cannot be selected', async() => {
    const onSelectMock = jest.fn()
    render(<Tree selectable={false} treeData={defaultTreeData} onSelect={onSelectMock} />)

    const [, vegetablesCaret] = screen.getAllByRole('img', { name: /caret-down/i })
    fireEvent.click(vegetablesCaret)

    const leekNode = await screen.findByText('Leek')
    fireEvent.click(leekNode)

    expect(onSelectMock).toHaveBeenCalledTimes(0)
  })

  test('nodes can be checked', async() => {
    const treeData: TreeDataNode[] = [
      {
        title: 'Fruits',
        key: 'fruits',
      },
      {
        title: 'Vegetables',
        key: 'vegetables',
      },
    ]

    const onCheckMock = jest.fn()
    const { container } = render(<Tree checkable={true} treeData={treeData} onCheck={onCheckMock} />)

    // Unfortunately, checkboxes inside the tree does not include the "checkbox" attribute
    // We need to find them via class name
    const checkboxes = container.getElementsByClassName('mia-platform-tree-checkbox')
    expect(checkboxes).toHaveLength(2)

    fireEvent.click(checkboxes[0])
    await waitFor(() => { expect(onCheckMock).toHaveBeenCalledTimes(1) })
    expect(onCheckMock).toHaveBeenCalledWith(['fruits'], expect.any(Object))

    fireEvent.click(checkboxes[1])
    await waitFor(() => { expect(onCheckMock).toHaveBeenCalledTimes(2) })
    expect(onCheckMock).toHaveBeenCalledWith(['fruits', 'vegetables'], expect.any(Object))

    fireEvent.click(checkboxes[0])
    await waitFor(() => { expect(onCheckMock).toHaveBeenCalledTimes(3) })
    expect(onCheckMock).toHaveBeenCalledWith(['vegetables'], expect.any(Object))
  })

  test('icons are visible', async() => {
    const treeData = [
      {
        title: 'Arrows',
        key: 'arrow',
        icon: <Icon name="PiArrowUpFill" size={16} />,
        children: [
          {
            title: 'Up',
            key: 'arrow-up',
            icon: <Icon name="PiArrowUp" size={16} />,
          },
          {
            title: 'Left',
            key: 'arrow-left',
            icon: <Icon name="PiArrowLeft" size={16} />,
          },
          {
            title: 'Down',
            key: 'arrow-down',
            icon: <Icon name="PiArrowDown" size={16} />,
          },
          {
            title: 'Right',
            key: 'arrow-right',
            icon: <Icon name="PiArrowRight" size={16} />,
          },
        ],
      },
    ]

    render(<Tree showIcon treeData={treeData} />)

    expect(screen.getByRole('img', { name: 'PiArrowUpFill' })).toBeInTheDocument()

    fireEvent.click(screen.getByRole('img', { name: /caret-down/i }))

    await waitFor(() => { expect(screen.getByRole('img', { name: 'PiArrowUp' })).toBeInTheDocument() })
    expect(screen.getByRole('img', { name: 'PiArrowDown' })).toBeInTheDocument()
    expect(screen.getByRole('img', { name: 'PiArrowLeft' })).toBeInTheDocument()
    expect(screen.getByRole('img', { name: 'PiArrowRight' })).toBeInTheDocument()
  })
})
