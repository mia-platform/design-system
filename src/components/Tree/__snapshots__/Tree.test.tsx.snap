// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree Component renders Tree 1`] = `
<DocumentFragment>
  <div
    class="mia-platform-tree mia-platform-tree-icon-hide customTree"
    role="tree"
  >
    <div>
      <input
        aria-label="for screen reader"
        style="width: 0px; height: 0px; display: flex; overflow: hidden; opacity: 0; border: 0px; padding: 0px; margin: 0px;"
        tabindex="0"
        value=""
      />
    </div>
    <div
      aria-hidden="true"
      class="mia-platform-tree-treenode"
      style="position: absolute; pointer-events: none; visibility: hidden; height: 0px; overflow: hidden; border: 0px; padding: 0px;"
    >
      <div
        class="mia-platform-tree-indent"
      >
        <div
          class="mia-platform-tree-indent-unit"
        />
      </div>
    </div>
    <div
      class="mia-platform-tree-list"
      style="position: relative;"
    >
      <div
        class="mia-platform-tree-list-holder"
      >
        <div>
          <div
            class="mia-platform-tree-list-holder-inner"
            style="display: flex; flex-direction: column;"
          >
            <div
              aria-grabbed="false"
              class="mia-platform-tree-treenode mia-platform-tree-treenode-switcher-close"
              draggable="false"
            >
              <span
                aria-hidden="true"
                class="mia-platform-tree-indent"
              />
              <span
                class="mia-platform-tree-switcher mia-platform-tree-switcher_close"
              >
                <span
                  aria-label="caret-down"
                  class="anticon anticon-caret-down mia-platform-tree-switcher-icon"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="caret-down"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="0 0 1024 1024"
                    width="1em"
                  >
                    <path
                      d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                    />
                  </svg>
                </span>
              </span>
              <span
                class="mia-platform-tree-node-content-wrapper mia-platform-tree-node-content-wrapper-close"
                title="Fruits"
              >
                <span
                  class="mia-platform-tree-title"
                >
                  Fruits
                </span>
              </span>
            </div>
            <div
              aria-grabbed="false"
              class="mia-platform-tree-treenode mia-platform-tree-treenode-switcher-close mia-platform-tree-treenode-leaf-last"
              draggable="false"
            >
              <span
                aria-hidden="true"
                class="mia-platform-tree-indent"
              />
              <span
                class="mia-platform-tree-switcher mia-platform-tree-switcher_close"
              >
                <span
                  aria-label="caret-down"
                  class="anticon anticon-caret-down mia-platform-tree-switcher-icon"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="caret-down"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="0 0 1024 1024"
                    width="1em"
                  >
                    <path
                      d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                    />
                  </svg>
                </span>
              </span>
              <span
                class="mia-platform-tree-node-content-wrapper mia-platform-tree-node-content-wrapper-close"
                title="Vegetables"
              >
                <span
                  class="mia-platform-tree-title"
                >
                  Vegetables
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
